### PR TITLE
(DRAFT): EG-710: Opening initial PR for feedback

### PR DESCRIPTION
**Pull request by rogersanick**
This PR includes an initial implementation / POC for a builder class to emulate Kotlin DSL for the Tokens SDK. Feedback is welcome! This is only an initial design proposal. 

See files 
- `workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt`
- `workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/TokenTypeBuilderTest.java`

### CODE REVIEW

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    private fun initializeAmountOrThrow(callbackSetter: () -> Unit): TokenBuilder {
+        if (!amountInitialized()) {
+            callbackSetter()
+        } else {
+            throw IllegalArgumentException("The token amount has already been initialized")
+        }
+        return this
+    }
+
+    fun amount(intAmount: Int) = initializeAmountOrThrow { this.intAmount = intAmount }
```
Comment (ivanr3) [inline workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt]: `ofAmount` or `withAmount` would probably be better.

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    fun amount(longAmount: Long) = initializeAmountOrThrow { this.longAmount = longAmount}
+    fun amount(doubleAmount: Double) = initializeAmountOrThrow { this.doubleAmount = doubleAmount }
+    fun amount(bigDecimalAmount: BigDecimal) = initializeAmountOrThrow { this.bigDecimalAmount = bigDecimalAmount }
+
+    fun <T: TokenType> of(t: T): TokenBuilder = when(getInitializedAmount()) {
```
Comment (ivanr3) [inline workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt]: What is this method for?

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import net.corda.core.identity.Party
+import java.lang.IllegalArgumentException
+import java.math.BigDecimal
+
+/**
+ * This is a utility class designed for Java developers to more easily
+ * access Kotlin infix functions to build token types.
+ */
+class TokenBuilder {
+    private var intAmount: Int? = null
```
Comment (ivanr3) [inline workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt]: Why do you need 4 different amounts here? Can you convert everything to long at the point of assignment?

File: workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/TokenTypeBuilderTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    /**
+     * Sanity check for easy access of digital currency TokenTypes from java code
+     */
+    @Test
+    public void javaWrappedDigitalCurrencyIsIdenticalToKotlinCompanionObject() throws Exception {
+        CordaX500Name aliceX500Name = new CordaX500Name("Alice", "NY", "US");
+        StartedMockNode alice = node(aliceX500Name);
+        Party aliceParty = alice.getInfo().getLegalIdentities().get(0);
+
+        new TokenBuilder()
```
Comment (ivanr3) [inline workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/TokenTypeBuilderTest.java]: How do you get a token from the builder?

Comment (anixon604) [reply workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/TokenTypeBuilderTest.java]: @rogersanick this ^^ was also my next question.

With the kot infix the primary return vals will be these cases:
        val token : TokenType = GBP
        val amountTokenType : Amount<TokenType> = 10.GBP
        val nonFunIssued : IssuedTokenType = GBP issuedBy party
        val funIssued : Amount<IssuedTokenType> = 10.GBP issuedBy party
        val nonFunIssuedAndHeld : NonFungibleToken = GBP issuedBy party heldBy party
        val funIssuedAndHeld : FungibleToken = 10.GBP issuedBy party heldBy party

With the general constraints
amount (optional) -> tokenType [R]-> issuedBy [R] -> heldBy [R]

Not sure if this is the BEST practice, but could have a closing .resolve() which could be inserted in the [R] points to return the given object.

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt (hunk 2 above)
Comment (rogersanick) [reply workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt]: This is a replication of the `of` method - which is an infix function applied to a number value (`Int`, `Long`, `BigDecimal`, `Double`) and accepting either a `TokenType` or a subtype of `TokenType` (e.g. - issued `IssuedTokenType`) hence the generics.

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt (hunk 3 above)
Comment (rogersanick) [reply workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt]: This is an interesting point, I was trying to keep this class as lightweight as possible. The conversion to `Long` is done in a very specific way in `AmountUtilities.kt` was hoping to leverage the functionality there so as to not duplicate.

File: workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/TokenTypeBuilderTest.java (hunk 4 above)
Comment (rogersanick) [reply workflows/src/test/kotlin/com/r3/corda/lib/tokens/workflows/TokenTypeBuilderTest.java]: ... my bad, this one should have been obvious. Hurray for POC's!

**Pull request closed without merging**